fem/qfunction.cpp

```diff
@@ -0,0 +1,265 @@
+// Copyright (c) 2010-2022, Lawrence Livermore National Security, LLC. Produced
+// at the Lawrence Livermore National Laboratory. All Rights reserved. See files
+// LICENSE and NOTICE for details. LLNL-CODE-806117.
+//
+// This file is part of the MFEM library. For more information and source code
+// availability visit https://mfem.org.
+//
+// MFEM is free software; you can redistribute it and/or modify it under the
+// terms of the BSD-3 license. We welcome feedback and contributions, see file
+// CONTRIBUTING.md for details.
+
+#include "qfunction.hpp"
+#include "quadinterpolator.hpp"
+#include "quadinterpolator_face.hpp"
+
+namespace mfem
+{
+
+QuadratureFunction &QuadratureFunction::operator=(double value)
+{
+   Vector::operator=(value);
+   return *this;
+}
+
+QuadratureFunction &QuadratureFunction::operator=(const Vector &v)
+{
+   MFEM_ASSERT(qspace && v.Size() == this->Size(), "");
+   Vector::operator=(v);
+   return *this;
+}
+
+
+QuadratureFunction::QuadratureFunction(Mesh *mesh, std::istream &in)
+{
+   const char *msg = "invalid input stream";
+   std::string ident;
+
+   qspace = new QuadratureSpace(mesh, in);
+   own_qspace = true;
+
+   in >> ident; MFEM_VERIFY(ident == "VDim:", msg);
+   in >> vdim;
+
+   Load(in, vdim*qspace->GetSize());
+}
+
+void QuadratureFunction::SetSpace(QuadratureSpaceBase *qspace_, int vdim_)
+{
+   if (qspace_ != qspace)
+   {
+      if (own_qspace) { delete qspace; }
+      qspace = qspace_;
+      own_qspace = false;
+   }
+   vdim = (vdim_ < 0) ? vdim : vdim_;
+   SetSize(vdim*qspace->GetSize());
+}
+
+void QuadratureFunction::SetSpace(
+   QuadratureSpaceBase *qspace_, double *qf_data, int vdim_)
+{
+   if (qspace_ != qspace)
+   {
+      if (own_qspace) { delete qspace; }
+      qspace = qspace_;
+      own_qspace = false;
+   }
+   vdim = (vdim_ < 0) ? vdim : vdim_;
+   NewDataAndSize(qf_data, vdim*qspace->GetSize());
+}
+
+void QuadratureFunction::Save(std::ostream &os) const
+{
+   GetSpace()->Save(os);
+   os << "VDim: " << vdim << '\n'
+      << '\n';
+   Vector::Print(os, vdim);
+   os.flush();
+}
+
+void QuadratureFunction::ProjectGridFunction(const GridFunction &gf)
+{
+   SetVDim(gf.VectorDim());
+
+   if (auto *qs_elem = dynamic_cast<QuadratureSpace*>(qspace))
+   {
+      const FiniteElementSpace &gf_fes = *gf.FESpace();
+      const bool use_tensor_products = UsesTensorBasis(gf_fes);
+      const ElementDofOrdering ordering = use_tensor_products ?
+                                          ElementDofOrdering::LEXICOGRAPHIC :
+                                          ElementDofOrdering::NATIVE;
+
+      // Use element restriction to go from L-vector to E-vector
+      const Operator *R = gf_fes.GetElementRestriction(ordering);
+      Vector e_vec(R->Height());
```

Comment:
For vectors like these, can we integrate to be able to specify the memory type to use (ex. Umpire temp memory pool) like the geometric factors in mesh.cpp

Reply:
That's a good idea. What would be the best interface to expose to the user? It would be a little weird to pass in a memory type, since this is just an internal temporary vector, it's more of an implementation detail than part of the public interface